### PR TITLE
Avoid warning when cloning tensor in self.log

### LINKED ISSUE #14594: UserWarning: use tensor.clone().detach() instead of torch.tensor(tensor)
## 🐛 Bug

This bug was introduced in Lightning 1.7.5 and caught by our CI when we tried to upgrade. We see the following warning message all of a sudden:
```
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:696: in fit
    self._call_and_handle_interrupt(
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:650: in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:735: in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:1166: in _run
    results = self._run_stage()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:1252: in _run_stage
    return self._run_train()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:1283: in _run_train
    self.fit_loop.run()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py:200: in run
    self.advance(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py:271: in advance
    self._outputs = self.epoch_loop.run(self._data_fetcher)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py:200: in run
    self.advance(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py:203: in advance
    batch_output = self.batch_loop.run(kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py:200: in run
    self.advance(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py:87: in advance
    outputs = self.optimizer_loop.run(optimizers, kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py:200: in run
    self.advance(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:201: in advance
    result = self._run_optimization(kwargs, self._optimizers[self.optim_progress.optimizer_position])
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:248: in _run_optimization
    self._optimizer_step(optimizer, opt_idx, kwargs.get("batch_idx", 0), closure)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:358: in _optimizer_step
    self.trainer._call_lightning_module_hook(
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:[155](https://github.com/microsoft/torchgeo/runs/8236896123?check_suite_focus=true#step:9:156)0: in _call_lightning_module_hook
    output = fn(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/core/module.py:[167](https://github.com/microsoft/torchgeo/runs/8236896123?check_suite_focus=true#step:9:168)2: in optimizer_step
    optimizer.step(closure=optimizer_closure)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/core/optimizer.py:[168](https://github.com/microsoft/torchgeo/runs/8236896123?check_suite_focus=true#step:9:169): in step
    step_output = self._strategy.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/strategies/strategy.py:216: in optimizer_step
    return self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py:153: in optimizer_step
    return optimizer.step(closure=closure, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/torch/optim/optimizer.py:113: in wrapper
    return func(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/torch/autograd/grad_mode.py:27: in decorate_context
    return func(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/torch/optim/adam.py:118: in step
    loss = closure()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py:138: in _wrap_closure
    closure_result = closure()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:146: in __call__
    self._result = self.closure(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:132: in closure
    step_output = self._step_fn()
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py:407: in _training_step
    training_step_output = self.trainer._call_strategy_hook("training_step", *kwargs.values())
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:[170](https://github.com/microsoft/torchgeo/runs/8236896123?check_suite_focus=true#step:9:171)4: in _call_strategy_hook
    output = fn(*args, **kwargs)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/strategies/strategy.py:358: in training_step
    return self.model.training_step(*args, **kwargs)
torchgeo/trainers/byol.py:430: in training_step
    self.log("train_loss", loss, on_step=True, on_epoch=False)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/core/module.py:422: in log
    value = apply_to_collection(value, (torch.Tensor, numbers.Number), self.__to_tensor, name)
/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytorch_lightning/utilities/apply_func.py:99: in apply_to_collection
    return function(data, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def __to_tensor(self, value: Union[torch.Tensor, numbers.Number], name: str) -> Tensor:
>       value = torch.tensor(value, device=self.device)
E       UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
```

### To Reproduce

This bug should be easy to see if you use pytest and convert all warnings to errors.

### Expected behavior

I would expect Lightning not to emit warnings about usage.

### Environment

- Lightning Component: LightningModule?
- PyTorch Lightning Version: 1.7.5
- Lightning App Version: N/A
- PyTorch Version: 1.12.1
- Python version: 3.10.6
- OS: Ubuntu 20.04.5
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A
- How you installed PyTorch: pip
- Running environment of LightningApp: CI
- Any other relevant information:

### Additional context

See https://github.com/microsoft/torchgeo/runs/8236896123?check_suite_focus=true for the CI failure.

cc @akihironitta

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #14594

Test fails on master. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -558,7 +558,7 @@ def __check_allowed(v: Any, name: str, value: Any) -> None:
         raise ValueError(f"`self.log({name}, {value})` was called, but `{type(v).__name__}` values cannot be logged")
 
     def __to_tensor(self, value: Union[torch.Tensor, numbers.Number], name: str) -> Tensor:
-        value = torch.tensor(value, device=self.device)
+        value = value.clone().detach() if isinstance(value, torch.Tensor) else torch.tensor(value, device=self.device)
```
Comment (awaelchli) [inline src/pytorch_lightning/core/module.py:561]: when it's a tensor, we need to move it to self.device too I guess.

**Comment by awaelchli** [comment]
@carmocca Since you are code-owner for the logging part, where would you suggest that I insert the test case?

File: tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py [hunk 2]
```diff
@@ -621,11 +622,26 @@ def training_step(self, *args):
 
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
     model = TestModel()
-    match = escape(f"self.log(foo, {value})` was called")
+    match = escape(f"self.log(foo, {value})` was called, but the tensor must have a single element")
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py:625]: note: it is hard to find tests corresponding to source code when only a tiny fraction of the error sentence is exposed in the tests.

Comment (carmocca) [reply tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py:625]: We expose a tiny fraction because we want to test only that the warning/error is shown, not the language used inside. It's inconvenient when you make a small grammatical change and then the tests fail. Fun fact, [expecttests](https://github.com/ezyang/expecttest) address this issue

File: tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py [hunk 3]
```diff
@@ -621,11 +622,26 @@ def training_step(self, *args):
...
+    """Regression test for issue https://github.com/Lightning-AI/lightning/issues/14594."""
+
+    class TestModel(BoringModel):
+        def training_step(self, *args):
+            self.log("foo", torch.tensor(1))
+            return super().training_step(*args)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    model = TestModel()
+    match = escape("To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach()")
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py:640]: I suggest making this text smaller so that a change in PyTorch to the text does not make this test useless.

```suggestion
    match = escape("recommended.*.clone().detach()")
```

(I didn't test if escape works with regex `.*`), might need tweaking

**Review by carmocca — COMMENTED** [review]
> @carmocca Since you are code-owner for the logging part, where would you suggest that I insert the test case?

The chosen file is appropriate

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py (hunk 3 above)
Comment (krshrimali) [reply tests/tests_pytorch/trainer/logging_/test_train_loop_logging.py:640]: @awaelchli - If you agree to this change, can you please commit? Or if it's resolved offline, please feel free to resolve this conversation. Let's try to get this PR to today's release 🚀

**Review by tchaton — APPROVED** [review]
LGTM !
